game-core/src/main/java/games/strategy/net/IpFinder.java

```diff
@@ -41,64 +45,26 @@
-    final Enumeration<NetworkInterface> enum1 = NetworkInterface.getNetworkInterfaces();
-    // Test if null, no point taking a performance hit by
-    // letting the JVM check for a NullPointerException.
-    if (enum1 == null) {
-      final InetAddress ip1 = InetAddress.getLocalHost();
-      return ip1;
-    }
-    final List<InetAddress> allButLoopback = new ArrayList<>();
-    while (enum1.hasMoreElements()) {
-      final NetworkInterface netface = enum1.nextElement();
-      final Enumeration<InetAddress> enum2 = netface.getInetAddresses();
-      while (enum2.hasMoreElements()) {
-        final InetAddress ip2 = enum2.nextElement();
-        if (!ip2.isLoopbackAddress()) {
-          allButLoopback.add(ip2);
-        }
-      }
-    }
-    // try to find one that is not private and ip4
-    for (final InetAddress address : allButLoopback) {
-      if (address.getAddress().length == 4 && isPublicNetworkAddress(address)) {
-        return address;
-      }
-    }
-    // try to find one that is not private
-    for (final InetAddress address : allButLoopback) {
-      if (isPublicNetworkAddress(address)) {
-        return address;
-      }
+    final Enumeration<NetworkInterface> interfaces = NetworkInterface.getNetworkInterfaces();
+    if (interfaces == null) {
+      return InetAddress.getLocalHost();
     }
-    // try to find one that is not link local
-    for (final InetAddress address : allButLoopback) {
-      if (!address.isLinkLocalAddress()) {
-        return address;
-      }
-    }
-    // all else fails, return localhost
-    return InetAddress.getLocalHost();
-  } // end static findInetAddress()
 
-  private static boolean isPublicNetworkAddress(final InetAddress address) {
-    // stupid java signed byte type
-    final byte octet254 = (byte) 0xFE;
-    final byte[] bytes = address.getAddress();
-    // ip 4
-    if (bytes.length == 4) {
-      // http://en.wikipedia.org/wiki/Private_network
-      final byte octet169 = (byte) 0xA9;
-      final byte octet168 = (byte) 0xA8;
-      final byte octet172 = (byte) 0xAC;
-      final byte octet192 = (byte) 0xC0;
-      return (bytes[0] != 10) && (bytes[0] != octet172 || bytes[1] < 16 || bytes[1] > 31)
-          && (bytes[0] != octet192 || bytes[1] != octet168) && (bytes[0] != octet169 || bytes[1] != octet254);
-    }
-    // ip 6
-    // http://en.wikipedia.org/wiki/IPv6#Addressing
-    final byte octet252 = (byte) 0xFC;
-    return (bytes[0] != octet252 || bytes[1] != 0) && bytes[0] != octet254;
+    return Collections.list(interfaces).stream()
+        .map(NetworkInterface::getInetAddresses)
+        .map(Collections::list)
+        .flatMap(Collection::stream)
+        .filter(Util.not(InetAddress::isLoopbackAddress))
+        .sorted(Comparator
+            .comparing(IpFinder::isPublic)
+            .thenComparing(Inet4Address.class::isInstance)
+            .thenComparing(InetAddress::isLinkLocalAddress, Comparator.reverseOrder())
+            .reversed())
```

Comment:
This always screws me up.  I had to write a few tests to verify it was doing what I thought it was. :smile:

```diff
diff --git a/game-core/src/main/java/games/strategy/net/IpFinder.java b/game-core/src/main/java/games/strategy/net/IpFinder.java
index c68922231..66edd7681 100644
--- a/game-core/src/main/java/games/strategy/net/IpFinder.java
+++ b/game-core/src/main/java/games/strategy/net/IpFinder.java
@@ -10,6 +10,8 @@ import java.util.Collections;
 import java.util.Comparator;
 import java.util.Enumeration;
 
+import com.google.common.annotations.VisibleForTesting;
+
 import games.strategy.util.Util;
 
 /**
@@ -55,15 +57,20 @@ public class IpFinder {
         .map(Collections::list)
         .flatMap(Collection::stream)
         .filter(Util.not(InetAddress::isLoopbackAddress))
-        .sorted(Comparator
-            .comparing(IpFinder::isPublic)
-            .thenComparing(Inet4Address.class::isInstance)
-            .thenComparing(InetAddress::isLinkLocalAddress, Comparator.reverseOrder())
-            .reversed())
+        .sorted(getInetAddressComparator())
         .findFirst()
         .orElse(InetAddress.getLocalHost());
   }
 
+  @VisibleForTesting
+  static Comparator<InetAddress> getInetAddressComparator() {
+    return Comparator
+        .comparing(IpFinder::isPublic)
+        .thenComparing(Inet4Address.class::isInstance)
+        .thenComparing(InetAddress::isLinkLocalAddress, Comparator.reverseOrder())
+        .reversed();
+  }
+
   private static boolean isPublic(final InetAddress address) {
     return !(address.isSiteLocalAddress() || address.isLinkLocalAddress());
   }
diff --git a/game-core/src/test/java/games/strategy/net/IpFinderTest.java b/game-core/src/test/java/games/strategy/net/IpFinderTest.java
new file mode 100644
index 000000000..adea36561
--- /dev/null
+++ b/game-core/src/test/java/games/strategy/net/IpFinderTest.java
@@ -0,0 +1,129 @@
+package games.strategy.net;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.stream.Stream;
+
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+
+public final class IpFinderTest {
+  // 172.217.3.132
+  private static final byte[] INET4_PUBLIC_ADDRESS = new byte[] {
+      (byte) 0xAC, (byte) 0xD9, (byte) 0x03, (byte) 0x84
+  };
+
+  // 192.168.2.1
+  private static final byte[] INET4_SITE_LOCAL_ADDRESS = new byte[] {
+      (byte) 0xC0, (byte) 0xA8, (byte) 0x02, (byte) 0x01
+  };
+
+  // 169.254.2.1
+  private static final byte[] INET4_LINK_LOCAL_ADDRESS = new byte[] {
+      (byte) 0xA9, (byte) 0xFE, (byte) 0x02, (byte) 0x01
+  };
+
+  // 2607:f8b0:4008:80c::2004
+  private static final byte[] INET6_PUBLIC_ADDRESS = new byte[] {
+      (byte) 0x26, (byte) 0x07, (byte) 0xF8, (byte) 0xB0, (byte) 0x40, (byte) 0x08, (byte) 0x08, (byte) 0x0C,
+      (byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x20, (byte) 0x04
+  };
+
+  // fec0::1ff:fe23:4567:890a
+  private static final byte[] INET6_SITE_LOCAL_ADDRESS = new byte[] {
+      (byte) 0xFE, (byte) 0xC0, (byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x00,
+      (byte) 0x01, (byte) 0xFF, (byte) 0xFE, (byte) 0x23, (byte) 0x45, (byte) 0x67, (byte) 0x89, (byte) 0x0A
+  };
+
+  // fe80::1ff:fe23:4567:890a
+  private static final byte[] INET6_LINK_LOCAL_ADDRESS = new byte[] {
+      (byte) 0xFE, (byte) 0x80, (byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x00,
+      (byte) 0x01, (byte) 0xFF, (byte) 0xFE, (byte) 0x23, (byte) 0x45, (byte) 0x67, (byte) 0x89, (byte) 0x0A
+  };
+
+  private static InetAddress inetAddressOf(final byte[] addr) {
+    try {
+      return InetAddress.getByAddress(addr);
+    } catch (final UnknownHostException e) {
+      throw new AssertionError("should not happen", e);
+    }
+  }
+
+  @BeforeAll
+  public static void checkInvariants() {
+    assertThat(inetAddressOf(INET4_LINK_LOCAL_ADDRESS).isLinkLocalAddress(), is(true));
+    assertThat(inetAddressOf(INET4_PUBLIC_ADDRESS).isLinkLocalAddress(), is(false));
+    assertThat(inetAddressOf(INET4_PUBLIC_ADDRESS).isSiteLocalAddress(), is(false));
+    assertThat(inetAddressOf(INET4_SITE_LOCAL_ADDRESS).isSiteLocalAddress(), is(true));
+
+    assertThat(inetAddressOf(INET6_LINK_LOCAL_ADDRESS).isLinkLocalAddress(), is(true));
+    assertThat(inetAddressOf(INET6_PUBLIC_ADDRESS).isLinkLocalAddress(), is(false));
+    assertThat(inetAddressOf(INET6_PUBLIC_ADDRESS).isSiteLocalAddress(), is(false));
+    assertThat(inetAddressOf(INET6_SITE_LOCAL_ADDRESS).isSiteLocalAddress(), is(true));
+  }
+
+  @Nested
+  public final class GetInetAddressComparatorTest {
+    private InetAddress selectInetAddress(final byte[]... addrs) {
+      return Stream.of(addrs)
+          .map(IpFinderTest::inetAddressOf)
+          .sorted(IpFinder.getInetAddressComparator())
+          .findFirst()
+          .orElseThrow(() -> new AssertionError("failed to select an InetAddress"));
+    }
+
+    @Test
+    public void shouldSelectInet4PublicAddressBeforeInet6PublicAddress() {
+      assertThat(
+          selectInetAddress(INET6_PUBLIC_ADDRESS, INET4_PUBLIC_ADDRESS),
+          is(inetAddressOf(INET4_PUBLIC_ADDRESS)));
+      assertThat(
+          selectInetAddress(INET4_PUBLIC_ADDRESS, INET6_PUBLIC_ADDRESS),
+          is(inetAddressOf(INET4_PUBLIC_ADDRESS)));
+    }
+
+    @Test
+    public void shouldSelectInet6PublicAddressBeforeInet4LinkLocalAddress() {
+      assertThat(
+          selectInetAddress(INET4_LINK_LOCAL_ADDRESS, INET6_PUBLIC_ADDRESS),
+          is(inetAddressOf(INET6_PUBLIC_ADDRESS)));
+      assertThat(
+          selectInetAddress(INET6_PUBLIC_ADDRESS, INET4_LINK_LOCAL_ADDRESS),
+          is(inetAddressOf(INET6_PUBLIC_ADDRESS)));
+    }
+
+    @Test
+    public void shouldSelectInet6PublicAddressBeforeInet4SiteLocalAddress() {
+      assertThat(
+          selectInetAddress(INET4_SITE_LOCAL_ADDRESS, INET6_PUBLIC_ADDRESS),
+          is(inetAddressOf(INET6_PUBLIC_ADDRESS)));
+      assertThat(
+          selectInetAddress(INET6_PUBLIC_ADDRESS, INET4_SITE_LOCAL_ADDRESS),
+          is(inetAddressOf(INET6_PUBLIC_ADDRESS)));
+    }
+
+    @Test
+    public void shouldSelectInet4SiteLocalAddressBeforeInet4LinkLocalAddress() {
+      assertThat(
+          selectInetAddress(INET4_LINK_LOCAL_ADDRESS, INET4_SITE_LOCAL_ADDRESS),
+          is(inetAddressOf(INET4_SITE_LOCAL_ADDRESS)));
+      assertThat(
+          selectInetAddress(INET4_SITE_LOCAL_ADDRESS, INET4_LINK_LOCAL_ADDRESS),
+          is(inetAddressOf(INET4_SITE_LOCAL_ADDRESS)));
+    }
+
+    @Test
+    public void shouldSelectInet6SiteLocalAddressBeforeInet6LinkLocalAddress() {
+      assertThat(
+          selectInetAddress(INET6_LINK_LOCAL_ADDRESS, INET6_SITE_LOCAL_ADDRESS),
+          is(inetAddressOf(INET6_SITE_LOCAL_ADDRESS)));
+      assertThat(
+          selectInetAddress(INET6_SITE_LOCAL_ADDRESS, INET6_LINK_LOCAL_ADDRESS),
+          is(inetAddressOf(INET6_SITE_LOCAL_ADDRESS)));
+    }
+  }
+}
```

Reply:
@ssoloff Had a hard time getting it to work as well ^^
I didn't write any tests for it, because I really think this class shouldn't be required, it should use a wildcard IP or a user-configured one.

Reply:
However I'd appreciate if you'd submit a PR with your testcases.